state/MMR.go

```diff
@@ -218,22 +317,218 @@ func (s *State) makeMMRs() {
+						// This is so the accounting adds upp.
+						s.MissingRequestAskCnt++
 					}
 					*when = now + askDelay // update when we asked
-
 					// Maybe when asking for past the end of the list we should not ask again?
 				}
 			} //build a MMRs with all the expired asks in that VM at that DBH.
 
 			for index, mmr := range mmrs {
 				s.LogMessage(logname, "sendout", mmr)
-				s.MissingRequestAskCnt++
 				if MMR_enable {
 					mmr.SendOut(s, mmr)
 				}
 				delete(mmrs, index)
 			} // Send MMRs that were built
-
-		}
+		} // select across all the channels, block till something happens
+		s.LogPrintf("mmr", "done")
 	} // forever ...
 } // func  makeMMRs() {...}
+
+// MissingMessageResponseCache will cache all processlist items from the last 2 blocks.
+// It can create MissingMessageResponses to peer requests, and prevent us from asking the network
+// if we already have something locally.
+type MissingMessageResponseCache struct {
+	// MissingMsgRequests is the channel on which we receive acked messages to cache
+	MissingMsgRequests chan interfaces.IMsg
+	// ProcessedPairs is all the ack+msg pairs that we processed
+	ProcessedPairs chan *MsgPair
+
+	// AckMessageCache is the cached acks from the last 2 blocks
+	AckMessageCache *AckMsgPairCache
+
+	// We need the state for getting the current timestamp and for logging
+	// TODO: Separate logging and current time from state
+	localState *State
+
+	quit chan bool
+}
+
+func NewMissingMessageReponseCache(s *State) *MissingMessageResponseCache {
+	mmrc := new(MissingMessageResponseCache)
+	mmrc.MissingMsgRequests = make(chan interfaces.IMsg, 20)
+	mmrc.ProcessedPairs = make(chan *MsgPair, 5)
+	mmrc.AckMessageCache = NewAckMsgCache()
+
+	mmrc.quit = make(chan bool, 1)
+	mmrc.localState = s
+
+	return mmrc
+}
+
+// NotifyPeerMissingMsg is the threadsafe way to notify that a peer sent us a missing message
+func (mmrc *MissingMessageResponseCache) NotifyPeerMissingMsg(missingMsg interfaces.IMsg) {
+	mmrc.MissingMsgRequests <- missingMsg
+}
+
+// NotifyNewMsgPair is the threadsafe way to include a new msg pair to respond to missing message requests
+// from peers
+func (mmrc *MissingMessageResponseCache) NotifyNewMsgPair(ack interfaces.IMsg, msg interfaces.IMsg) {
+	mmrc.ProcessedPairs <- &MsgPair{Ack: ack, Msg: msg}
+}
+
+func (mmrc *MissingMessageResponseCache) Close() {
+	mmrc.quit <- true
+}
+
+// answerRequest will attempt to construct a missing message response for a missing message request.
+// 		params:
+//			request
+func (mmrc *MissingMessageResponseCache) answerRequest(request *messages.MissingMsg) {
+	// Loop through all requested heights
+	for _, plHeight := range request.ProcessListHeight {
+		// Check if we have the response for a given plHeight
+		pair := mmrc.AckMessageCache.Get(int(request.DBHeight), request.VMIndex, int(plHeight))
+		if pair != nil { // Woo! Respond to that peer
+			if pair.Msg == nil || pair.Ack == nil {
+				panic("This should never happen")
```

Comment:
probably log and survive this in case it happens in deployment.. really we need a better way to assert in debug and survive in production that we always use in case like this.